src/github.com/getlantern/enproxy/conn_test.go

```diff
@@ -62,18 +60,80 @@ func TestBadBuffered(t *testing.T) {
 	doTestBad(true, t)
 }
 
+func TestIdle(t *testing.T) {
+	idleTimeout := 100 * time.Millisecond
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	_, err = Dial(httpAddr, &Config{
+		DialProxy: func(addr string) (net.Conn, error) {
+			return net.Dial("tcp", proxyAddr)
+		},
+		NewRequest:  newRequest,
+		IdleTimeout: idleTimeout,
+	})
+	if assert.NoError(t, err, "Dialing should have succeeded") {
+		time.Sleep(idleTimeout * 2)
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+	}
+}
+
+// This test stimulates a connection leak as seen in
+// https://github.com/getlantern/lantern/issues/2174.
+func TestHTTPRedirect(t *testing.T) {
+	startProxy(t)
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(network, addr string) (net.Conn, error) {
+				return Dial(addr, &Config{
+					DialProxy: func(addr string) (net.Conn, error) {
+						return net.Dial("tcp", proxyAddr)
+					},
+					NewRequest: newRequest,
+				})
+			},
+			DisableKeepAlives: true,
+		},
+	}
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	resp, err := client.Head("http://www.facebook.com")
+	if assert.NoError(t, err, "Head request to facebook should have succeeded") {
+		resp.Body.Close()
+	}
+
+	assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+}
+
 func doTestPlainText(buffered bool, t *testing.T) {
 	startServers(t)
 
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
 	conn, err := prepareConn(httpAddr, buffered, false, t)
 	if err != nil {
 		t.Fatalf("Unable to prepareConn: %s", err)
 	}
-	defer conn.Close()
+	defer func() {
+		err := conn.Close()
+		assert.Nil(t, err, "Closing conn should succeed")
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
```

Comment:
Seems looks like we have an odd race condition here but I haven't figure out the cause. I found it while running `go tests` for the first time, but it didn't happend the second time until a ran `go test` inside a while loop.

```
--- FAIL: TestPlainTextStreaming (1.10s)
        Location:       conn_test.go:131
        Error:          No error is expected but got Unexpected TCP file descriptor count. Expected 2, have 0.

                        New file descriptors
                        -----------------------------
                        enproxy.t 22767  rev   10u     IPv4 511988      0t0     TCP localhost.localdomain:50181->localhost.localdomain:58955 (ESTABLISHED)
                        enproxy.t 22767  rev    9u     IPv4 511987      0t0     TCP localhost.localdomain:58955->localhost.localdomain:50181 (ESTABLISHED)

                        Removed file descriptors
                        -----------------------------
                        enproxy.t 22767  rev    7u     IPv4 512067      0t0     TCP localhost.localdomain:43616->localhost.localdomain:57883 (CLOSE_WAIT)
                        enproxy.t 22767  rev    9u     IPv4 510929      0t0     TCP localhost.localdomain:39772->localhost.localdomain:46442 (CLOSE_WAIT)
        Messages:       All file descriptors except the connection from proxy to destination site should have been closed
```

 The `CLOSE_WAIT` status may have something to do with it? or a waiting time somewhere? I'll investigate more on the topic.

I'm triggering the race condition and capturing the output with:

```
while true; do go test &>> log.txt; sleep 1; done
```

It eventually happens.

Any clues where should I start looking?